runtime/tests/port/j9strTest.c

```diff
@@ -1528,46 +1528,58 @@ j9str_WinacpToMutf8(struct J9PortLibrary *portLibrary)
 #if defined(J9ZOS390)
 /**
  * Check whether a null/non-null terminated string is properly handled by fstring().
+ * Also checks if number, long number, long long number, and float are properly handled.
  * It calls atoe_vsnprintf() to verify the functionalities.
  * @param[in] portLibrary - The port library under test
  * @param[in] precision - flag for precision specifier
  * @param[in] buffer - pointer to the buffer intended for atoe_vsnprintf()
  * @param[in] bufferLength - the buffer length
  * @param[in] expectedOutput - The expected string
+ * @param[in] expectedCount - The expected response from atoe_vsnprintf()
  * @param[in] format - The string format
  *
  * @return TEST_PASS on success
  *         TEST_FAIL on error
  */
 static int
-printToBuffer(struct J9PortLibrary *portLibrary, BOOLEAN precision, char *buffer, size_t bufferLength, const char *expectedOutput, const char *format, ...)
+printToBuffer(struct J9PortLibrary *portLibrary, BOOLEAN precision, char *buffer, size_t bufferLength, const char *expectedOutput, int expectedCount, const char *format, ...)
 {
 	PORT_ACCESS_FROM_PORT(portLibrary);
+	static const char NULLCHARSTRING[] = "[null]";
 	int stringLength = 0;
+	int expectedLength = 0;
+	int actualCount = 0;
+	char *nonNullExpectedOutput;
+	char *nonNullBuffer;
 	va_list args;
 
 	va_start(args, format);
-	stringLength = atoe_vsnprintf(buffer, bufferLength, format, args);
-	va_end (args);
+	actualCount = atoe_vsnprintf(buffer, bufferLength, format, args);
+	expectedLength = (NULL == expectedOutput) ? 0 : strlen(expectedOutput);
+	stringLength = (NULL == buffer) ? 0 : strlen(buffer);
+	nonNullExpectedOutput = (NULL == expectedOutput) ? NULLCHARSTRING : expectedOutput;
+	nonNullBuffer = (NULL == buffer) ? NULLCHARSTRING : buffer;
+	va_end(args);
 
-	if (stringLength >= 0) {
+	if (actualCount >= 0) {
 		if (precision) {
-			j9tty_printf(PORTLIB, "\n\tFinish Testing: String in buffer is: \"%s\", length = %d\n", buffer, stringLength);
+			j9tty_printf(PORTLIB, "\n\tFinish Testing: String in buffer is: \"%s\", length = %d\n", nonNullBuffer, stringLength);
 		} else {
-			j9tty_printf(PORTLIB, "\n\tFinish Testing without precision specifier: String in buffer is: \"%s\", length = %d\n", buffer, stringLength);
+			j9tty_printf(PORTLIB, "\n\tFinish Testing without precision specifier: String in buffer is: \"%s\", length = %d\n", nonNullBuffer, stringLength);
 		}
 
-		if ((stringLength == strlen(expectedOutput))
-			&& (0 == memcmp(buffer, expectedOutput, stringLength))
+		if ((stringLength == expectedLength)
+			&& (expectedCount == actualCount)
+			&& (0 == memcmp(nonNullBuffer, nonNullExpectedOutput, stringLength))
```

Comment:
This is not safe (if either refers to `NULLCHARSTRING` and `stringLength > strlen(NULLCHARSTRING)`).

Reply:
Suggest:
```c
			&& ((NULL == buffer) || (0 == memcmp(buffer, expectedOutput, expectedCount + 1)))
```
This handles the tests where we're only verifying the length returned by `atoe_vsnprintf()` and checks that any output is null-terminated.

Reply:
> This is not safe (if either refers to `NULLCHARSTRING` and `stringLength > strlen(NULLCHARSTRING)`).

in this case the first condition `if ((stringLength == expectedLength)` would fail since the length is set to zero for the null string and would be non zero for the other string.

Reply:
currently the only way it could pass is if both strings are equal length and value. the strlen would take care of null termination. We can not use the `expectedCount` since it could be much higher than the length of strings. I could use 
```
			&& ((NULL == buffer) || (0 == memcmp(buffer, expectedOutput, stringLength + 1)))
```

Reply:
I updated it to `&& (0 == memcmp(nonNullBuffer, nonNullExpectedOutput, stringLength + 1))` this way, it would compare at least one element and termination chars. in case one string is empty and the other is NULL, it would compare first char of `"" vs "[null]"` and since '\0' != '[', it fails